c7n/resources/vpc.py

```diff
@@ -3272,3 +3272,140 @@ def process(self, resources, event=None):
                 results.append(resource)
 
         return results
+
+
+@Vpc.filter_registry.register('resolver-rules-associated')
+class VpcResolverRulesAssociatedFilter(BaseAction):
+    """Filter VPCs which have either missing or matching groups
+    of Route53 Resolver Rules and annotate for action.
+
+    :Example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: vpcs-missing-resolver-rules
+                resource: vpc
+                filters:
+                  - type: resolver-rules-associated
+                    name: my-rule.*
+                    associated: False
+    """
+    schema = type_schema(
+        'resolver-rules-associated',
+        name={'type': 'string'},
+        associated={'type': 'boolean'},
+        required=['name', 'associated']
+    )
+    permissions = (
+        'route53resolver:ListResolverRules',
+        'route53resolver:ListResolverRuleAssociations'
+    )
+
+    def get_all(self, client, type, key):
+        paginator = client.get_paginator(type)
+        paginator.PAGE_ITERATOR_CLS = query.RetryPageIterator
+        results = paginator.paginate().build_full_result().get(key, [])
+        return results
+
+    def process(self, resources, event=None):
+        results = []
+        target_rule_name = self.data['name']
+        associated = self.data['associated']
+        client = local_session(self.manager.session_factory).client('route53resolver')
+
+        rules = self.get_all(client, 'list_resolver_rules', 'ResolverRules')
+        rules_set = set()
+        for rule in rules:
+            if re.match(target_rule_name, rule['Name']):
+                rules_set.add(rule['Id'])
+
+        rule_associations = self.get_all(client,
+            'list_resolver_rule_associations',
+            'ResolverRuleAssociations'
+        )
+        vpc_rule_map = {}
+        for a in rule_associations:
+            if a['VPCId'] in vpc_rule_map:
+                vpc_rule_map[a['VPCId']] += [a['ResolverRuleId']]
+            else:
+                vpc_rule_map[a['VPCId']] = [a['ResolverRuleId']]
```

Comment:
nit: we should be able to avoid a level of nesting here

```suggestion
            vpc_rule_map.setdefault(a['VPCId'], []).append(a['ResolverRuleId'])
```